### ISSUE TITLE
Is there any plan to expose intermediate CNN outputs in hub with KerasLayer?

### ISSUE THREAD

**Issue by jiahuei** [tensorflow/hub#453]
Currently all CNNs in Hub only provide either feature vector after global pooling or logits.

However it can be very useful to have access to the feature maps as well as retaining ability to fine-tune the CNN using KerasLayer in TF 2.0.

Is there any plan to expose intermediate layer outputs in hub with KerasLayer?

**Comment by gowthamkpr** [comment]
Do you have any specific module in mind @jiahuei ?

**Comment by jiahuei** [comment]
Hi there,

Mainly CNNs that are unavailable via `tf.keras.applications`, such as
InceptionV1 and InceptionV4.

Thanks.

Best regards,

Tan Jia Huei
University of Malaya.

**Comment by arnoegw** [comment]
Hi @jiahuei, there are no current plans to expose intermediate outputs systematically.

However, there is an undocumented way to get them out of some TF2 SavedModels exported from TF-Slim, such as https://tfhub.dev/google/imagenet/inception_v1/feature_vector/4: passing `return_endpoints=True` to the SavedModel's `__call__` function changes the output to a dict.

**NOTE:** This interface is subject to change or removal, and has known issues.

```python
l = hub.KerasLayer(
    "https://tfhub.dev/google/imagenet/inception_v1/feature_vector/4",
    trainable=True,  # Or not, as you please.
    arguments=dict(return_endpoints=True))
images = tf.keras.layers.Input((224, 224, 3))
outputs = l(images)
for k, v in sorted(outputs.items()):
  print(k, v.shape)
```

Issues to be aware of:
  * Undocumented, subject to change or removal, not available consistently.
  * `__call__` computes all outputs (and applies all update ops during training) irrespective of the ones being used later on.

**Comment by agonojo** [comment]
jiahuei have you since been able to easily do this with models such as efficientnet? It really is annoying that we can't easily see intermediate layers. Even when I've tried to do it by utilizing tf.keras_applications models to just transfer over my weights, differences in the architectures make it nearly impossible to do even though documentation states "they are identical."

**Comment by jiahuei** [comment]
Hi, it has been a while, but I might be able to retrieve intermediate layers. Check out the code at https://github.com/jiahuei/TF2-pretrained-CNN

**Comment by egormkn** [comment]
Hi @arnoegw! Are there any changes expected to support access to intermediate layers? I've found a lot of issues in this repository about accessing intermediate outputs for BERT and various CNN models, and this issue seems to be the only one where a solution is given. Please, consider implementing some kind of official way to use intermediate outputs, as it will significantly expand the capabilities of tfhub models.

**Comment by ujjwal-ai** [comment]
Is this issue receiving any attention ? This is one of the most important concerns right now.

**Comment by akhorlin** [comment]
Currently, this issue is not under active development. However given that
in TF any tensor can be retrieved using tf.get_variable, one somewhat hacky
approach could be to inspect a graph (say in tensorboard.dev), find the
name of relevant tensors and read them directly

**Comment by matfeb** [comment]
Hi, 

in order to see your model architecture, you can try (example with efficientnet lite0):

```python
efficientnet_lite0_base_layer = hub.KerasLayer(
    "https://tfhub.dev/tensorflow/efficientnet/lite0/feature-vector/2",
    output_shape=[1280],
    trainable=False
)

print("Thickness of the model:", len(efficientnet_lite0_base_layer.weights))
print ("{:<80} {:<20} {:<10}".format('Layer','Shape','Type'))

for i in range(len(efficientnet_lite0_base_layer.weights)):
    model_weights_raw_string = str(efficientnet_lite0_base_layer.weights[i])
    model_weights_wo_weights = model_weights_raw_string.split(", numpy", 1)[0]
    dtype = model_weights_wo_weights.split(" dtype=")[1]
    shape = model_weights_wo_weights.split(" shape=")[1].split(" dtype=")[0]
    
    print ("{:<80} {:<20} {:<10}".format(efficientnet_lite0_base_layer.weights[i].name, shape, dtype))
```
Output:
```
Thickness of the model: 245
Layer                                                                            Shape                Type      
efficientnet-lite0/stem/conv2d/kernel:0                                          (3, 3, 3, 32)        float32   
efficientnet-lite0/stem/tpu_batch_normalization/gamma:0                          (32,)                float32   
efficientnet-lite0/stem/tpu_batch_normalization/beta:0                           (32,)                float32   
efficientnet-lite0/stem/tpu_batch_normalization/moving_mean:0                    (32,)                float32   
efficientnet-lite0/stem/tpu_batch_normalization/moving_variance:0                (32,)                float32   
efficientnet-lite0/blocks_0/depthwise_conv2d/depthwise_kernel:0                  (3, 3, 32, 1)        float32   
efficientnet-lite0/blocks_0/tpu_batch_normalization/gamma:0                      (32,)                float32   
efficientnet-lite0/blocks_0/tpu_batch_normalization/beta:0                       (32,)                float32   
efficientnet-lite0/blocks_0/tpu_batch_normalization/moving_mean:0                (32,)                float32   
efficientnet-lite0/blocks_0/tpu_batch_normalization/moving_variance:0            (32,)                float32   
efficientnet-lite0/blocks_0/conv2d/kernel:0                                      (1, 1, 32, 16)       float32   
efficientnet-lite0/blocks_0/tpu_batch_normalization_1/gamma:0                    (16,)                float32   
efficientnet-lite0/blocks_0/tpu_batch_normalization_1/beta:0                     (16,)                float32   
efficientnet-lite0/blocks_0/tpu_batch_normalization_1/moving_mean:0              (16,)                float32   
efficientnet-lite0/blocks_0/tpu_batch_normalization_1/moving_variance:0          (16,)                float32   
.
.
.
efficientnet-lite0/blocks_15/conv2d/kernel:0                                     (1, 1, 192, 1152)    float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization/gamma:0                     (1152,)              float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization/beta:0                      (1152,)              float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization/moving_mean:0               (1152,)              float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization/moving_variance:0           (1152,)              float32   
efficientnet-lite0/blocks_15/depthwise_conv2d/depthwise_kernel:0                 (3, 3, 1152, 1)      float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization_1/gamma:0                   (1152,)              float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization_1/beta:0                    (1152,)              float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization_1/moving_mean:0             (1152,)              float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization_1/moving_variance:0         (1152,)              float32   
efficientnet-lite0/blocks_15/conv2d_1/kernel:0                                   (1, 1, 1152, 320)    float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization_2/gamma:0                   (320,)               float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization_2/beta:0                    (320,)               float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization_2/moving_mean:0             (320,)               float32   
efficientnet-lite0/blocks_15/tpu_batch_normalization_2/moving_variance:0         (320,)               float32   
efficientnet-lite0/head/conv2d/kernel:0                                          (1, 1, 320, 1280)    float32   
efficientnet-lite0/head/tpu_batch_normalization/gamma:0                          (1280,)              float32   
efficientnet-lite0/head/tpu_batch_normalization/beta:0                           (1280,)              float32   
efficientnet-lite0/head/tpu_batch_normalization/moving_mean:0                    (1280,)              float32   
efficientnet-lite0/head/tpu_batch_normalization/moving_variance:0                (1280,)              float32   
```

**Comment by nss-ysasaki** [comment]
@akhorlin Hmmm..., I am wondering how I can get *intermediate layer outputs* from `tf.Variable`s. Could you elaborate a bit?
